drivers/adc/adc_npcx.c

```diff
@@ -0,0 +1,374 @@
+/*
+ * Copyright (c) 2020 Nuvoton Technology Corporation.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT nuvoton_npcx_adc
+
+#include <assert.h>
+#include <drivers/adc.h>
+#include <drivers/clock_control.h>
+#include <kernel.h>
+#include <soc.h>
+
+#define ADC_CONTEXT_USES_KERNEL_TIMER
+#include "adc_context.h"
+
+#include <logging/log.h>
+LOG_MODULE_REGISTER(adc_npcx, CONFIG_ADC_LOG_LEVEL);
+
+/* ADC speed/delay values during initialization */
+#define ADC_REGULAR_DLY_VAL     0x03
+#define ADC_REGULAR_ADCCNF2_VAL 0x8B07
+#define ADC_REGULAR_GENDLY_VAL  0x0100
+#define ADC_REGULAR_MEAST_VAL   0x0001
+
+/* ADC channel number */
+#define NPCX_ADC_CH_COUNT 10
+
+/* ADC targeted operating frequency (2MHz) */
+#define NPCX_ADC_CLK 2000000
+
+/* ADC internal reference voltage (Unit:mV) */
+#define NPCX_ADC_VREF_VOL 2816
+
+/* ADC conversion mode */
+#define NPCX_ADC_CHN_CONVERSION_MODE  0
+#define NPCX_ADC_SCAN_CONVERSION_MODE 1
+
+/* Device config */
+struct adc_npcx_config {
+	/* adc controller base address */
+	uintptr_t base;
+	/* clock configuration */
+	struct npcx_clk_cfg clk_cfg;
+	/* pinmux configuration */
+	const struct npcx_alt *alts_list;
+};
+
+/* Driver data */
+struct adc_npcx_data {
+	/* Input clock for ADC converter */
+	uint32_t input_clk;
+	/* mutex of ADC channels */
+	struct adc_context ctx;
+	/*
+	 * Bit-mask indicating the channels to be included in each sampling
+	 * of this sequence.
+	 */
+	uint16_t channels;
+	/* Device pointer to get HAL instance */
+	const struct device *adc_dev;
+	uint16_t *buffer;
+	uint16_t *repeat_buffer;
+};
+
+/* Driver convenience defines */
+#define DRV_CONFIG(dev) \
+	((const struct adc_npcx_config *)(dev)->config)
+
+#define DRV_DATA(dev) \
+	((struct adc_npcx_data *)(dev)->data)
+
+#define HAL_INSTANCE(dev) \
+	(struct adc_reg *)(DRV_CONFIG(dev)->base)
+
+/* ADC local functions */
+static void adc_npcx_isr(void *arg)
+{
+	struct adc_npcx_data *const data = DRV_DATA((const struct device *)arg);
+	struct adc_reg *const inst = HAL_INSTANCE((const struct device *)arg);
+	uint16_t status = inst->ADCSTS;
+	uint16_t result, channel;
+
+	/* Clear status pending bits first */
+	inst->ADCSTS = status;
+	LOG_DBG("%s: status is %04X\n", __func__, status);
+
+	/* Is end of conversion cycke event? ie. Scan conversion is done. */
+	if (IS_BIT_SET(status, NPCX_ADCSTS_EOCCEV)) {
+		/* Stop conversion for scan conversion mode */
+		inst->ADCCNF |= BIT(NPCX_ADCCNF_STOP);
+
+		/* Get result for each ADC selected channel */
+		while (data->channels) {
+			channel = find_lsb_set(data->channels) - 1;
+			result = GET_FIELD(inst->CHNDAT[channel],
+					NPCX_CHNDAT_CHDAT_FIELD);
+			/* Save ADC result */
+			*data->buffer++ = result;
```

Comment:
What protection do we have here for buffer overflow? (We need to add code protection or comments)

Reply:
That's a little difficult to check the length of data->buffer here. Per my understanding, it has been validated by adc_npcx_validate_buffer_size(). I don't see the other ADC drivers have the similar implementation.

Reply:
We can add the comment that adc_npcx_validate_buffer_size ensures that the buffer here will have enough space for the ISR to write.

We could also store a `size_t bytes_left` along with buffer in the data struct. We would set it to the size in adc_npcx_validate_buffer_size and decrement it here each time we wrote to it. Then we could know that we never wrote past valid memory and it only take an extra 4 bytes of RAM for the check (seems worth doing to me)

Reply:
Done.